### PR TITLE
ClickHouse: Normalize one batch at a time

### PR DESCRIPTION
**Pull request by Amogh-Bharadwaj**
This helps with:
- Memory utilisation on ClickHouse side
- Observability
- Debuggability

### CODE REVIEW

File: flow/connectors/clickhouse/normalize.go [hunk 1]
```diff
@@ -266,19 +267,84 @@ func (c *ClickHouseConnector) NormalizeRecords(
...
+		`SELECT DISTINCT _peerdb_destination_table_name FROM %s WHERE _peerdb_batch_id = %d`,
+		rawTbl, batchID)
+
+	rows, err := c.query(ctx, q)
+	if err != nil {
+		return nil, fmt.Errorf("error while querying raw table for distinct table names in batch: %w", err)
+	}
+	defer rows.Close()
+	var tableNames []string
+	for rows.Next() {
```
Comment (serprex) [inline flow/connectors/clickhouse/normalize.go:312]: Use pgx.CollectRows

Comment (heavycrystal) [reply flow/connectors/clickhouse/normalize.go:312]: this is a query on ClickHouse connector, don't think that's possible

File: flow/connectors/clickhouse/normalize.go [hunk 2]
```diff
@@ -266,19 +267,84 @@ func (c *ClickHouseConnector) NormalizeRecords(
...
+
+func (c *ClickHouseConnector) getDistinctTableNamesInBatch(
+	ctx context.Context,
+	flowJobName string,
+	batchID int64,
+) ([]string, error) {
+	rawTbl := c.getRawTableName(flowJobName)
+
+	q := fmt.Sprintf(
+		`SELECT DISTINCT _peerdb_destination_table_name FROM %s WHERE _peerdb_batch_id = %d`,
```
Comment (heavycrystal) [inline flow/connectors/clickhouse/normalize.go:303]: is it possible to craft a query to aggregate and fetch distinct tables for all batches at once

File: flow/connectors/clickhouse/normalize.go [hunk 3]
```diff
@@ -266,19 +267,84 @@ func (c *ClickHouseConnector) NormalizeRecords(
...
 		c.logger.Error("[clickhouse] error while getting distinct table names in batch", "error", err)
-		return nil, err
+		return err
 	}
-
-	rawTbl := c.getRawTableName(req.FlowJobName)
-
 	// model the raw table data as inserts.
 	for _, tbl := range destinationTableNames {
 		// SELECT projection FROM raw_table WHERE _peerdb_batch_id > normalize_batch_id AND _peerdb_batch_id <= sync_batch_id
```
Comment (serprex) [inline flow/connectors/clickhouse/normalize.go:350]: outdated comment

File: flow/connectors/clickhouse/normalize.go (hunk 2 above)
Comment (Amogh-Bharadwaj) [reply flow/connectors/clickhouse/normalize.go:303]: That is the current behaviour and this PR changes it to get it for each batch
Do you want this function to be called higher up for all batches and pass a map of batchId:[table1,table2..] down to `syncTablesInThisBatch()`

Comment (heavycrystal) [reply flow/connectors/clickhouse/normalize.go:303]: yes

File: flow/connectors/clickhouse/normalize.go [hunk 4]
```diff
@@ -266,19 +267,84 @@ func (c *ClickHouseConnector) NormalizeRecords(
 		return nil, fmt.Errorf("failed to copy avro stages to destination: %w", err)
 	}
 
+	rawTbl := c.getRawTableName(req.FlowJobName)
+	for batchID := normBatchID + 1; batchID <= req.SyncBatchID; batchID++ {
+		err = c.syncTablesInThisBatch(ctx, req, rawTbl, batchID)
+		if err != nil {
```
Comment (serprex) [inline flow/connectors/clickhouse/normalize.go:273]:
```suggestion
		if err := c.syncTablesInThisBatch(ctx, req, rawTbl, batchID); err != nil {
```

**Review by serprex — APPROVED** [review]
followup should migrate to 64bit batch ids

**Pull request merged**